orchagent/debug_counter/drop_counter.cpp

```diff
@@ -0,0 +1,298 @@
+#include "drop_counter.h"
+
+#include "logger.h"
+#include "sai_serialize.h"
+
+using std::runtime_error;
+using std::string;
+using std::unordered_map;
+using std::unordered_set;
+using std::vector;
+
+extern sai_object_id_t gSwitchId;
+extern sai_debug_counter_api_t *sai_debug_counter_api;
+
+const int maxDropReasons = 100;
+
+// If initialization fails, this constructor will throw a runtime error.
+DropCounter::DropCounter(const string& counter_name, const string& counter_type, const unordered_set<string>& drop_reasons)
+        : DebugCounter(counter_name, counter_type), drop_reasons(drop_reasons)
+{
+    SWSS_LOG_ENTER();
+    this->initializeDropCounterInSAI();
+}
+
+DropCounter::~DropCounter()
+{
+    SWSS_LOG_ENTER();
+    try
+    {
+        DebugCounter::removeDebugCounterFromSAI();
+    }
+    catch (const std::exception& e)
+    {
+        SWSS_LOG_ERROR("Failed to remove drop counter '%s' from SAI", this->name.c_str());
+    }
+}
+
+// If we are unable to query the SAI or the type of counter is not supported
+// then this method throws a runtime error.
+std::string DropCounter::getDebugCounterSAIStat() const
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t index_attribute;
+    index_attribute.id = SAI_DEBUG_COUNTER_ATTR_INDEX;
+    if (sai_debug_counter_api->get_debug_counter_attribute(this->counter_id, 1, &index_attribute) != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to get stat for debug counter '%s'", this->name.c_str());
+        throw runtime_error("Failed to get debug counter stat");
+    }
+
+    auto index = index_attribute.value.u32;
```

Comment:
How do we know that index is not out of range?

Reply:
I'm not sure that I understand the feedback here. `index` is just the value returned from the call to the SAI.

Reply:
The index is used in all the if blocks below. How do we know that the index will come out valid after (BASE+index) operation?

Reply:
The indices are managed by the SAI. By the time we reach this step, the counter has already been initialized, so we can assume the index is valid. However, there is a limit to the number of enums that are defined in the SAI. I don't believe any of the vendors support enough counters for this to be an issue, but I added a check just to save someone a lot of confusion in the future. :)